File: inventory-iter.go
package main

import (
	"context"
	"encoding/csv"
	"encoding/json"
	"encoding/xml"
	"flag"
	"fmt"
	"log"
	"os"
	"strings"
	"time"

	"go.etcd.io/etcd/client/v3"
)

// KeyValue represents a key-value pair.
type KeyValue struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

func connectToEtcd(etcdHost string, etcdPort int) (*clientv3.Client, error) {
	client, err := clientv3.New(clientv3.Config{
		Endpoints:   []string{fmt.Sprintf("%s:%d", etcdHost, etcdPort)},
		DialTimeout: 5 * time.Second,
	})
	if err != nil {
		return nil, err
	}
	return client, nil
}

func iterateEtcdKeys(client *clientv3.Client, keyPrefixes []string) <-chan KeyValue {
	keyValues := make(chan KeyValue)
	ctx, cancel := context.WithCancel(context.Background())

	go func() {
		defer cancel()
		defer close(keyValues)

		for _, keyPrefix := range keyPrefixes {
			resp, err := client.Get(ctx, keyPrefix, clientv3.WithPrefix())
			if err != nil {
				log.Printf("Failed to iterate over etcd keys: %v\n", err)
				return
			}

			for _, kv := range resp.Kvs {
				keyValues <- KeyValue{Key: string(kv.Key), Value: string(kv.Value)}
			}
		}
	}()

	return keyValues
}

func outputCSV(keyValues <-chan KeyValue) {
	csvWriter := csv.NewWriter(os.Stdout)
	defer csvWriter.Flush()

	for kv := range keyValues {
		if err := csvWriter.Write([]string{kv.Key, kv.Value}); err != nil {
			log.Fatalf("Error writing CSV: %v\n", err)
		}
	}
}

func outputTable(keyValues <-chan KeyValue) {
	fmt.Println("Key\tValue")
	for kv := range keyValues {
		fmt.Printf("%s\t%s\n", kv.Key, kv.Value)
	}
}

func outputJSON(keyValues <-chan KeyValue) {
	keyValueSlice := []KeyValue{}
	for kv := range keyValues {
		keyValueSlice = append(keyValueSlice, kv)
	}

	jsonBytes, err := json.MarshalIndent(keyValueSlice, "", "    ")
	if err != nil {
		log.Fatalf("Error marshaling JSON: %v\n", err)
	}

	fmt.Println(string(jsonBytes))
}

func outputXML(keyValues <-chan KeyValue) {
	keyValueSlice := []KeyValue{}
	for kv := range keyValues {
		keyValueSlice = append(keyValueSlice, kv)
	}

	output := struct {
		XMLName xml.Name `xml:"key_values"`
		Data    []KeyValue
	}{Data: keyValueSlice}

	xmlBytes, err := xml.MarshalIndent(output, "", "    ")
	if err != nil {
		log.Fatalf("Error marshaling XML: %v\n", err)
	}

	fmt.Println(xml.Header + string(xmlBytes))
}

func main() {
	etcdHost := flag.String("etcd-host", "", "etcd server address")
	etcdPort := flag.Int("etcd-port", 0, "etcd server port")
	keyPrefixes := flag.String("key-prefixes", "", "List of key prefixes to filter (comma-separated)")
	outputFormat := flag.String("output", "table", "Output format (csv, table, json, xml)")

	flag.Parse()

	if *etcdHost == "" || *etcdPort == 0 || *keyPrefixes == "" {
		log.Fatal("etcd-host, etcd-port, and key-prefixes are required")
	}

	client, err := connectToEtcd(*etcdHost, *etcdPort)
	if err != nil {
		log.Fatalf("Failed to connect to etcd: %v\n", err)
	}
	defer client.Close()

	keyPrefixList := strings.Split(*keyPrefixes, ",")
	keyValues := iterateEtcdKeys(client, keyPrefixList)

	switch *outputFormat {
	case "csv":
		outputCSV(keyValues)
	case "table":
		outputTable(keyValues)
	case "json":
		outputJSON(keyValues)
	case "xml":
		outputXML(keyValues)
	default:
		log.Fatalf("Invalid output format: %s\n", *outputFormat)
	}
}

File: python/inventory-iter.py
#!/usr/bin/env python3
import etcd3
import argparse
import json
import tabulate
import xml.etree.ElementTree as ET


def connect_to_etcd(etcd_host, etcd_port):
    try:
        client = etcd3.client(host=etcd_host, port=etcd_port)
        return client
    except Exception as e:
        print(f"Failed to connect to etcd: {str(e)}")
        return None


def iterate_etcd_keys(client, key_prefixes):
    try:
        for key_prefix in key_prefixes:
            for key, value in client.get_prefix(key_prefix):
                yield (key.decode("utf-8"), value.decode("utf-8"))
    except Exception as e:
        print(f"Failed to iterate over etcd keys: {str(e)}")


def output_csv(key_values):
    for key, value in key_values:
        print(f"{key}, {value}")


def output_table(key_values):
    if key_values:
        headers = ["Key", "Value"]
        data = list(key_values)
        print(tabulate.tabulate(data, headers=headers, tablefmt="grid"))
    else:
        print("No key-value pairs found in etcd.")


def output_json(key_values):
    key_value_dict = dict(key_values)
    print(json.dumps(key_value_dict, indent=4))


def output_xml(key_values):
    root = ET.Element("key_values")
    for key, value in key_values:
        key_value_element = ET.SubElement(root, "key_value")
        ET.SubElement(key_value_element, "key").text = key
        ET.SubElement(key_value_element, "value").text = value
    xml_string = ET.tostring(root, encoding="unicode")
    print(xml_string)


def output(hosts, output_format):
    output_functions = {
        "csv": output_csv,
        "table": output_table,
        "json": output_json,
        "xml": output_xml,
    }


def main():
    parser = argparse.ArgumentParser(description="Iterate over key-value pairs in etcd and output their contents.")
    parser.add_argument("--etcd-host", required=True, help="etcd server address")
    parser.add_argument("--etcd-port", type=int, required=True, help="etcd server port")
    parser.add_argument("--key-prefixes", required=True, nargs="+", help="List of key prefixes to filter")
    parser.add_argument("--output", choices=["csv", "table", "json", "xml"], default="table", help="Output format")

    args = parser.parse_args()

    if client := connect_to_etcd(args.etcd_host, args.etcd_port):
        try:
            key_values = iterate_etcd_keys(client, args.key_prefixes)
            output(key_values, args.output)
        except Exception as e:
            print(f"An error occurred: {str(e)}")


if __name__ == "__main__":
    main()
